include/nanobind/stl/detail/nb_set.h

```diff
@@ -0,0 +1,56 @@
+#pragma once
+
+#include <nanobind/nanobind.h>
+
+NAMESPACE_BEGIN(NB_NAMESPACE)
+NAMESPACE_BEGIN(detail)
+
+template <typename Value_, typename Key> struct set_caster {
+    NB_TYPE_CASTER(Value_, const_name("set[") + make_caster<Key>::Name + const_name("]"));
+
+    using KeyCaster = make_caster<Key>;
+
+    bool from_python(handle src, uint8_t flags, cleanup_list *cleanup) noexcept {
+        value.clear();
+
+        PyObject* iter = obj_iter(src.ptr());
+        if (iter == nullptr) {
+            PyErr_Clear();
+            return false;
+        }
+
+        Py_ssize_t size = NB_SET_GET_SIZE(src.ptr());
+        bool success = (size >= 0);
+
+        KeyCaster key_caster;
+        while (PyObject* key = obj_iter_next(iter)) {
```

Comment:
potential issue: `obj_iter_next` can raise, but `from_python` cannot. We need a non-throwing version of that function which indicates a failure (and clears the underlying Python error status) so that the caster can return `false` in this case.

Reply:
minor: is in't it common to put an extra pair of parentheses in a `while` statement when a variable is defined there? Perhaps it's cleaner to use a `for` statement (without upper bound) that has that idea builtin.

Reply:
Thanks! I've changed this to a for-loop. And the error has been checked if `key` is a `nullptr` (exit the for-loop), where we will print & clear the error.